kwok/operator/operator.go

```diff
@@ -0,0 +1,312 @@
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instance"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instanceprofile"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instancetype"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/launchtemplate"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/pricing"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/securitygroup"
+	ssmp "github.com/aws/karpenter-provider-aws/pkg/providers/ssm"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/subnet"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/version"
+	"github.com/aws/karpenter-provider-aws/pkg/utils"
+)
+
+func init() {
+	karpv1.NormalizedLabels = lo.Assign(karpv1.NormalizedLabels, map[string]string{"topology.ebs.csi.aws.com/zone": corev1.LabelTopologyZone})
+}
+
+// Operator is injected into the AWS CloudProvider's factories
+type Operator struct {
+	*operator.Operator
+	Config                      aws.Config
+	UnavailableOfferingsCache   *awscache.UnavailableOfferings
+	SSMCache                    *cache.Cache
+	ValidationCache             *cache.Cache
+	SubnetProvider              subnet.Provider
+	SecurityGroupProvider       securitygroup.Provider
+	InstanceProfileProvider     instanceprofile.Provider
+	AMIProvider                 amifamily.Provider
+	AMIResolver                 amifamily.Resolver
+	LaunchTemplateProvider      launchtemplate.Provider
+	PricingProvider             pricing.Provider
+	VersionProvider             *version.DefaultProvider
+	InstanceTypesProvider       *instancetype.DefaultProvider
+	InstanceProvider            instance.Provider
+	SSMProvider                 ssmp.Provider
+	CapacityReservationProvider capacityreservation.Provider
+	EC2API                      *kwokec2.Client
+}
+
+func NewOperator(ctx context.Context, operator *operator.Operator) (context.Context, *Operator) {
+	cfg := prometheusv2.WithPrometheusMetrics(WithUserAgent(lo.Must(config.LoadDefaultConfig(ctx))), crmetrics.Registry)
+	cfg.APIOptions = append(cfg.APIOptions, middleware.StructuredErrorHandler)
+	if cfg.Region == "" {
+		log.FromContext(ctx).V(1).Info("retrieving region from IMDS")
+		region := lo.Must(imds.NewFromConfig(cfg).GetRegion(ctx, nil))
+		cfg.Region = region.Region
+	}
+	ec2api := kwokec2.NewClient(cfg.Region, ec2.NewFromConfig(cfg), kwokec2.NewNopRateLimiterProvider(), strategy.NewLowestPrice(pricing.NewAPI(cfg), ec2.NewFromConfig(cfg), cfg.Region), operator.GetClient(), operator.Clock, operator.GetConfig())
+	go func() {
+		<-operator.Elected()
+		ec2api.StartBackupThread(ctx)
+	}()
+	go func() {
+		<-operator.Elected()
+		ec2api.StartKillNodeThread(ctx)
+	}()
+
+	eksapi := eks.NewFromConfig(cfg)
+	log.FromContext(ctx).WithValues("region", cfg.Region).V(1).Info("discovered region")
+	clusterEndpoint, err := ResolveClusterEndpoint(ctx, eksapi)
+	if err != nil {
+		log.FromContext(ctx).Error(err, "failed detecting cluster endpoint")
+		os.Exit(1)
```

Comment:
Any clean up required for the Backup/Kill threads?

Reply:
That's a good point -- we should wait for the goroutines to finish so that we make sure we don't kill things in the middle of writing to ConfigMaps